### PR TITLE
Adding backface culling to stairs

### PR DESCRIPTION
**Pull request by MrRar**
I used this Node.js script:

var fs = require("fs")
var file = fs.readFileSync("init.lua").toString()
var newFile = file
var matches = []
var cursor = 0;
while(true) {
	var pngPtr = file.indexOf(".png", cursor)
	if(pngPtr == -1) break
	matches.push(pngPtr)
	cursor = pngPtr + 4
}
for(var i in matches) {
	var end = matches[i]
	var start = file.lastIndexOf("\"", end)
	++start
	//console.log(start + ":" + end)
	var imgName = file.slice(start, end)
	var oldTexture = '"' + imgName + '.png"'
	var newTexture = "{name = \"" + imgName + ".png\", backface_culling = true}"
	newFile = newFile.replace(oldTexture, newTexture)
}
fs.writeFileSync("init.lua",newFile)

### CODE REVIEW

**Comment by tenplus1** [comment]
What advantages does it give?

**Comment by paramat** [comment]
Possibly faster render since it does not render the faces facing away from the camera?

**Comment by paramat** [comment]
screenshot_20170608_115153

^ Inside a stair node.

Works.
Interestingly a slab already had backface culling somehow, maybe because it is a nodeboxes converted to a meshnode, instead of an .obj meshnodes like the stair.

**Comment by paramat** [comment]
:+1:

**Comment by sfan5** [comment]
couldn't `stairs.register_stair_and_slab` just set `backface_culling` in the nodedef before registering?

**Comment by paramat** [comment]
Good point, once at line 54.

**Comment by MrRar** [comment]
@sfan5 That assumes that no stair will require backface rendering. Is that a good assumption to make? 
Should another parameter be added to register_stair_and_slab for whether to have backface culling for the stair?

**Comment by paramat** [comment]
> Is that a good assumption to make?

Erm, some mods may be registering glass stairs, do they need backfaces rendered?

**Comment by paramat** [comment]
> That assumes that no stair will require backface rendering. Is that a good assumption to make?

This is fine, because slabs (that arendeboxes converted to meshnodes) automatically do not have backface culling (and never have had), so stairs just become consistent with slabs. Glass slabs and stairs do not need backface rendering.

**Comment by paramat** [comment]
Thanks will test.

**Comment by paramat** [comment]
How about
```
	for index, image in pairs(images) do
		images[index] = {
			name = image,
			backface_culling = true,
		}
	end
 	groups.stair = 1
 	minetest.register_node(":stairs:stair_" .. subname, {
 		description = description,
 		drawtype = "mesh",
 		mesh = "stairs_stair.obj",
		tiles = images,
```

**Comment by paramat** [comment]
DS-minetest also warns that 'images' can now only contain a .png file, no other fields.

**Comment by DS-Minetest** [comment]
Actually I would rather suggest something like:
```
for i = 1, #images do
	if type(images[i]) == "string" then
		images[i] = {
			name = images[i],
			backface_culling = true,
		}
	else
		images[i].backface_culling = images[i].backface_culling == nil and true or images[i].backface_culling
	end
end
```
But I'm not sure.

**Comment by sfan5** [comment]
That code will modify the passed table, use something like this:
```lua
for i, image in ipairs(images) do
	if type(image) == "string" then
		images[i] = {
			name = image,
			backface_culling = true,
		}
	else
		if image.backface_culling == nil then -- override using any other value
			images[i] = table.copy(image)
			images[i].backface_culling = true
		end
	end
end
```

**Comment by MrRar** [comment]
@sfan5 It does not work because you're still modifying the original images table.

**Comment by paramat** [comment]
Latest version seems potentially good.

**Comment by sfan5** [comment]
oh yeah, well just throw a `images = table.copy(images)` in there

File: mods/stairs/init.lua [hunk 1]
```diff
@@ -46,12 +46,24 @@ end
...
 function stairs.register_stair(subname, recipeitem, groups, images, description, sounds)
+	local stair_images = {}
+	for i, image in ipairs(images) do
+		if type(image) == "string" then
+			stair_images[i] = {
+				name = image,
+				backface_culling = true,
+			}
+		elseif image.backface_culling == nil then
+			stair_images[i] = table.copy(image)
```
Comment (paramat) [inline mods/stairs/init.lua]: `stair_images = table.copy(images)`?

Comment (paramat) [inline mods/stairs/init.lua]: Is this correct? or
`stair_images = table.copy(images)`?
or
`stair_images[i] = images[i]`?

File: mods/stairs/init.lua [hunk 2]
```diff
@@ -46,12 +46,24 @@ end
...
 
 function stairs.register_stair(subname, recipeitem, groups, images, description, sounds)
+	local stair_images = {}
+	for i, image in ipairs(images) do
+		if type(image) == "string" then
+			stair_images[i] = {
+				name = image,
+				backface_culling = true,
+			}
+		elseif image.backface_culling == nil then
```
Comment (sfan5) [inline mods/stairs/init.lua]: you should keep the comment about overriding here

File: mods/stairs/init.lua (hunk 1 above)
Comment (sfan5) [inline mods/stairs/init.lua]: Neither of those are correct, it already does what it's supposed to.

Comment (paramat) [inline mods/stairs/init.lua]: Ok, but your comment suggests there is an error.
 > oh yeah, well just throw a `images = table.copy(images)` in there

Comment (sfan5) [inline mods/stairs/init.lua]: Yes, the original images table should either be copied or (like here) constructed from scratch.

Comment (paramat) [inline mods/stairs/init.lua]: Ok i see now, seems good as it is.

**Comment by paramat** [comment]
Code seems good now.

**Comment by paramat** [comment]
I'm now thinking it's better to just add 'backface_culling = true' to the tiles as the PR originally did, because that teaches modders to add that by being visible in the registrations. It's better that modders are aware of the advantage instead of hding it and doing it automatically for them.

**Comment by sfan5** [comment]
Adding a note to the docs should suffice IMO.

**Comment by paramat** [comment]
+1 if the author has tested.

File: mods/stairs/init.lua [hunk 3]
```diff
@@ -46,12 +46,24 @@ end
...
 
 function stairs.register_stair(subname, recipeitem, groups, images, description, sounds)
+	local stair_images = {}
+	for i, image in ipairs(images) do
+		if type(image) == "string" then
+			stair_images[i] = {
+				name = image,
+				backface_culling = true,
+			}
+		elseif image.backface_culling == nil then -- override using any other value
```
Comment (paramat) [inline mods/stairs/init.lua]: `elseif image.backface_culling != true then`
To override someone setting it to 'false'?

Comment (sfan5) [inline mods/stairs/init.lua]: (in that case the if condition would be useless anyway)

why?
maybe a modder wants to set it to false for some reason, that should be supported

Comment (paramat) [inline mods/stairs/init.lua]: Ok, i don't mind, but the comment suggests otherwise:
` -- override using any other value`
FIne as is then.

Comment (sfan5) [inline mods/stairs/init.lua]: the comment is a little ambiguous, it's supposed to mean that "the code that sets it to true can be overriden (disabled) by using (setting) any other value (than `nil`) in the definition table"

Comment (paramat) [inline mods/stairs/init.lua]: I see, fine.

**Comment by paramat** [comment]
:+1:

**Comment by DS-Minetest** [comment]
This needs documentation.

**Comment by paramat** [comment]
I don't think docs in game_api.txt are needed.

**Comment by sfan5** [comment]
Tested and works. :+1:

**Comment by paramat** [comment]
Applying: Adding backface culling to stairs
error: patch failed: mods/stairs/init.lua:342
error: mods/stairs/init.lua: patch does not apply

**Comment by sfan5** [comment]
You need to use `git am -3`

**Comment by paramat** [comment]
Ok 3 way merge.

**Comment by paramat** [comment]
7b46d3b6afa7c14a6ed26368820334b1adfc5e4d

**Pull request closed without merging**